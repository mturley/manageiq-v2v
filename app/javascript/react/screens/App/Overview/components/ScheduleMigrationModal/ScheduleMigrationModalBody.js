import React from 'react';
import { Icon } from 'patternfly-react';
import PropTypes from 'prop-types';
import { initializeDatepicker } from '../../helpers';

class ScheduleMigrationModalBody extends React.Component {
  componentDidMount() {
    if (this.props.showDatepicker) {
      const { handleDatepickerChange, defaultDate } = this.props;
      initializeDatepicker(handleDatepickerChange, defaultDate);
    }
  }

  componentDidUpdate(prevProps, prevState, snapshot) {
    if (prevProps.showDatepicker !== this.props.showDatepicker) {
      if (this.props.showDatepicker) {
        const { handleDatepickerChange, defaultDate } = this.props;
        initializeDatepicker(handleDatepickerChange, defaultDate);
      }
    }
  }

  render() {
    const { setScheduleMode, startNowLabel, startLaterLabel, showDatepicker } = this.props;

    const handleRadioChange = event => {
      setScheduleMode(event.currentTarget.value === 'schedule_migration_now');
    };

    return (
      <div>
        <div className="row">
          <div className="col-xs-12">
            <div className="form-group">
              <div className="radio">
                <label style={{ fontWeight: '600' }}>
                  <input
                    type="radio"
                    name="schedule_migration_start"
                    value="schedule_migration_now"
                    onChange={handleRadioChange}
                    checked={!showDatepicker}
                  />
                  {startNowLabel}
                </label>
              </div>
              <div className="radio">
                <label style={{ fontWeight: '600' }}>
                  <input
                    type="radio"
                    name="schedule_migration_start"
                    value="schedule_migration_later"
                    onChange={handleRadioChange}
                    checked={showDatepicker}
                  />
                  {startLaterLabel}
                </label>
              </div>
            </div>
          </div>
        </div>
        {showDatepicker && (
          <div className="row">
            <div className="col-xs-12">
              <div className="form-group">
                <div id="dateTimePicker" className="input-group date-time-picker-pf">
                  <input id="dateTimeInput" type="text" className="form-control" />
                  <span className="input-group-addon">
                    <Icon type="fa" name="calendar" />
                  </span>
                </div>
              </div>
            </div>
          </div>
        )}
      </div>
    );
  }
}

ScheduleMigrationModalBody.propTypes = {
  handleDatepickerChange: PropTypes.func,
  setScheduleMode: PropTypes.func,
  defaultDate: PropTypes.instanceOf(Date),
  startNowLabel: PropTypes.string,
  startLaterLabel: PropTypes.string,
  showDatepicker: PropTypes.bool
};

export default ScheduleMigrationModalBody;
