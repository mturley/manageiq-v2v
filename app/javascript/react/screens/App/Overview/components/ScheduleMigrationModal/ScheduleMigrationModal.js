import React from 'react';
import PropTypes from 'prop-types';
import { Button, Modal } from 'patternfly-react';
import ScheduleMigrationModalBody from './ScheduleMigrationModalBody';
import getPlanScheduleInfo from '../Migrations/helpers/getPlanScheduleInfo';

class ScheduleMigrationModal extends React.Component {
  constructor(props) {
    super(props);

    const { scheduleMigrationPlan } = this.props;
    const { migrationScheduled } = getPlanScheduleInfo(scheduleMigrationPlan) || '';

    this.state = {
      dateTimeInput: migrationScheduled ? new Date(migrationScheduled) : null,
      startMigrationNow: null
    };
  }

  render() {
    const {
      toggleScheduleMigrationModal,
      scheduleMigrationModal,
      scheduleMigrationPlan,
      scheduleMigration,
      fetchTransformationPlansAction,
      fetchTransformationPlansUrl,
      migrateClick
    } = this.props;

    const handleDatepickerChange = event => {
      this.setState({ dateTimeInput: event });
    };

    const startMigrationNowHandler = event => {
      this.setState({ startMigrationNow: event });
    };

    const modalClose = () => {
      toggleScheduleMigrationModal();
    };

    const modalTitle = plan =>
      plan != null && (plan.status === 'Error' || plan.status === 'Denied')
        ? __('Schedule Retry')
        : __('Schedule Migration');

    return (
      <Modal show={scheduleMigrationModal} onHide={modalClose} backdrop="static">
        <Modal.Header>
          <Modal.CloseButton onClick={modalClose} />
          <Modal.Title>{modalTitle(scheduleMigrationPlan)}</Modal.Title>
        </Modal.Header>
        <Modal.Body>
          <ScheduleMigrationModalBody
            handleDatepickerChange={handleDatepickerChange}
            startMigrationNowHandler={startMigrationNowHandler}
            defaultDate={this.state.dateTimeInput}
          />
        </Modal.Body>
        <Modal.Footer>
          <Button bsStyle="default" className="btn-cancel" onClick={modalClose}>
            {__('Cancel')}
          </Button>
          <Button
            bsStyle="primary"
            onClick={() => {
              if (this.state.startMigrationNow) {
                migrateClick(scheduleMigrationPlan.href);
                toggleScheduleMigrationModal();
              } else {
                scheduleMigration({
                  plan: scheduleMigrationPlan,
                  scheduleTime: this.state.dateTimeInput
                }).then(() => {
                  fetchTransformationPlansAction({
                    url: fetchTransformationPlansUrl,
                    archived: false
                  });
                });
              }
            }}
          >
            {__('Schedule')}
          </Button>
        </Modal.Footer>
      </Modal>
    );
  }
}

ScheduleMigrationModal.propTypes = {
  scheduleMigrationModal: PropTypes.bool,
  toggleScheduleMigrationModal: PropTypes.func,
  scheduleMigrationPlan: PropTypes.object,
  scheduleMigration: PropTypes.func,
  migrateClick: PropTypes.func,
  fetchTransformationPlansAction: PropTypes.func,
  fetchTransformationPlansUrl: PropTypes.string
};

export default ScheduleMigrationModal;
