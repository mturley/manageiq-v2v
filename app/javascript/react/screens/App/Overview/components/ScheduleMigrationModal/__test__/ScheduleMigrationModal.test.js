import React from 'react';
import { shallow } from 'enzyme';
import ScheduleMigrationModal from '../ScheduleMigrationModal';

describe('When initialized with warm migration plan', () => {
  const plan = {
    status: 'Ok',
    options: { config_info: { warm_migration: true } }
  };
  const baseProps = {
    scheduleMigrationModal: true,
    toggleScheduleMigrationModal: jest.fn(() => Promise.resolve()),
    scheduleMigrationPlan: plan,
    scheduleMigration: jest.fn(() => Promise.resolve()),
    scheduleMigrationNow: jest.fn(() => Promise.resolve()),
    scheduleCutover: jest.fn(() => Promise.resolve()),
    fetchTransformationPlansAction: jest.fn(() => Promise.resolve()),
    fetchTransformationPlansUrl: 'http://some-url/'
  };

  test('renders schedule cutover modal', () => {
    const wrapper = shallow(<ScheduleMigrationModal {...baseProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('modal buttons work correctly', async () => {
    const wrapper = shallow(<ScheduleMigrationModal {...baseProps} />);
    wrapper.find('Button[bsStyle="primary"]').simulate('click');
    expect(baseProps.scheduleCutover).toHaveBeenCalled();
    wrapper.find('Button[bsStyle="default"]').simulate('click');
    expect(baseProps.toggleScheduleMigrationModal).toHaveBeenCalled();
  });
});

describe('When initialized with cold migration plan', () => {
  const plan = {
    status: 'Ok',
    options: { config_info: { warm_migration: false } }
  };
  const baseProps = {
    scheduleMigrationModal: true,
    toggleScheduleMigrationModal: jest.fn(() => Promise.resolve()),
    scheduleMigrationPlan: plan,
    scheduleMigration: jest.fn(() => Promise.resolve()),
    scheduleMigrationNow: jest.fn(() => Promise.resolve()),
    scheduleCutover: jest.fn(() => Promise.resolve()),
    fetchTransformationPlansAction: jest.fn(() => Promise.resolve()),
    fetchTransformationPlansUrl: 'http://some-url/'
  };

  test('renders schedule migration modal', () => {
    const wrapper = shallow(<ScheduleMigrationModal {...baseProps} />);
    expect(wrapper).toMatchSnapshot();
  });

  test('modal buttons work correctly', async () => {
    const wrapper = shallow(<ScheduleMigrationModal {...baseProps} />);
    wrapper.find('Button[bsStyle="primary"]').simulate('click');
    expect(baseProps.scheduleMigration).toHaveBeenCalled();
    wrapper.find('Button[bsStyle="default"]').simulate('click');
    expect(baseProps.toggleScheduleMigrationModal).toHaveBeenCalled();
  });
});

describe('When initialized with failed cold migration plan', () => {
  const plan = {
    status: 'Error',
    options: { config_info: { warm_migration: false } }
  };
  const baseProps = {
    scheduleMigrationModal: true,
    toggleScheduleMigrationModal: jest.fn(() => Promise.resolve()),
    scheduleMigrationPlan: plan,
    scheduleMigration: jest.fn(() => Promise.resolve()),
    scheduleMigrationNow: jest.fn(() => Promise.resolve()),
    scheduleCutover: jest.fn(() => Promise.resolve()),
    fetchTransformationPlansAction: jest.fn(() => Promise.resolve()),
    fetchTransformationPlansUrl: 'http://some-url/'
  };

  test('renders schedule retry modal', () => {
    const wrapper = shallow(<ScheduleMigrationModal {...baseProps} />);
    expect(wrapper).toMatchSnapshot();
  });
});
