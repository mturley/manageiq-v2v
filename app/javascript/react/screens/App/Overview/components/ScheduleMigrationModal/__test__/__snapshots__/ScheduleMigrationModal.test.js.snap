// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`When initialized with cold migration plan renders schedule migration modal 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop="static"
  bsClass="modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={false}
    closeLabel="Close"
  >
    <ModalCloseButton
      className=""
      closeText="Close"
      onClick={[Function]}
    />
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Schedule Migration
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <ScheduleMigrationModalBody
      defaultDate={1970-01-01T00:00:00.000Z}
      handleDatepickerChange={[Function]}
      setScheduleMode={[Function]}
      showDatepicker={true}
      startLaterLabel="Select date and time for the start of the migration"
      startNowLabel="Start migration immediately"
    />
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      className="btn-cancel"
      disabled={false}
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Schedule
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`When initialized with failed cold migration plan renders schedule retry modal 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop="static"
  bsClass="modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={false}
    closeLabel="Close"
  >
    <ModalCloseButton
      className=""
      closeText="Close"
      onClick={[Function]}
    />
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Schedule Retry
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <ScheduleMigrationModalBody
      defaultDate={1970-01-01T00:00:00.000Z}
      handleDatepickerChange={[Function]}
      setScheduleMode={[Function]}
      showDatepicker={true}
      startLaterLabel="Select date and time to retry migration"
      startNowLabel="Retry migration immediately"
    />
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      className="btn-cancel"
      disabled={false}
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Schedule
    </Button>
  </ModalFooter>
</Modal>
`;

exports[`When initialized with warm migration plan renders schedule cutover modal 1`] = `
<Modal
  animation={true}
  autoFocus={true}
  backdrop="static"
  bsClass="modal"
  dialogComponentClass={[Function]}
  enforceFocus={true}
  keyboard={true}
  manager={
    ModalManager {
      "add": [Function],
      "containers": Array [],
      "data": Array [],
      "handleContainerOverflow": true,
      "hideSiblingNodes": true,
      "isTopModal": [Function],
      "modals": Array [],
      "remove": [Function],
    }
  }
  onHide={[Function]}
  renderBackdrop={[Function]}
  restoreFocus={true}
  show={true}
>
  <ModalHeader
    bsClass="modal-header"
    closeButton={false}
    closeLabel="Close"
  >
    <ModalCloseButton
      className=""
      closeText="Close"
      onClick={[Function]}
    />
    <ModalTitle
      bsClass="modal-title"
      componentClass="h4"
    >
      Schedule Cutover
    </ModalTitle>
  </ModalHeader>
  <ModalBody
    bsClass="modal-body"
    componentClass="div"
  >
    <ScheduleMigrationModalBody
      defaultDate={1970-01-01T00:00:00.000Z}
      handleDatepickerChange={[Function]}
      setScheduleMode={[Function]}
      showDatepicker={true}
      startLaterLabel="Select date and time to start cutover"
      startNowLabel="Start cutover immediately"
    />
  </ModalBody>
  <ModalFooter
    bsClass="modal-footer"
    componentClass="div"
  >
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="default"
      className="btn-cancel"
      disabled={false}
      onClick={[Function]}
    >
      Cancel
    </Button>
    <Button
      active={false}
      block={false}
      bsClass="btn"
      bsStyle="primary"
      disabled={false}
      onClick={[Function]}
    >
      Schedule
    </Button>
  </ModalFooter>
</Modal>
`;
